### ISSUE TITLE
Use Prettier

### ISSUE THREAD

**Issue by wchargin** [tensorflow/tensorboard#2466]
@stephanwlee and I want to run [Prettier][prettier] on TensorBoard.

Prettier is a code formatter for JavaScript, TypeScript, HTML, and CSS
(among others). Its approach is different than that of most formatting
linters, like ESLint: instead of identifying “bad patterns” and
suggesting peephole fixes, it simply parses the entire file into a
concrete syntax tree and re-emits the tree with a pretty-printing
algorithm. This “uncompromising” nature makes it a pleasure to work
with: it’s pretty much entirely insensitive to however the input was
already formatted, and will give you entirely predictable results.

(It’s like Go’s `gofmt`, or Python’s [`black`][black].)

Anecdotally, I noticed that my process of writing code changed fairly
significantly when using Prettier. I can write hideous-looking one-line
monsters, then simply save the file and have it automatically formatted
correctly. (And if it doesn’t reformat on save then I know that I have a
syntax error!) Unlike with ESLint and friends, I can _actually_
completely forget about the formatting.

## Why?

Most importantly, you don’t have to think about formatting while
writing, reviewing, or reading code.

Consistent style also makes diffs smaller, and improves greppability
(e.g., you don’t have to account for two quote forms).

### Style

Prettier’s code style uses a bit more whitespace than our existing code
style. Running Prettier on our codebase increases the total LOC in JS,
TS, HTML, and CSS files by 18%.

Prettier also indents TypeScript namespaces, while we historically have
not, so this adds ~1 level of indentation to all TypeScript files.

This is all totally fine, imho; just wanted to mention it.

### One-shot versus per-file incremental

The diff is fairly large:

```
$ git ls-files -z '*.js' '*.ts' '*.html' '*.css' | xargs -0 wc -l | tail -n 1
  87771 total
$ prettier --write 'tensorboard/**/*.'{js,ts,html,css} >/dev/null 2>/dev/null
$ git ls-files -z '*.js' '*.ts' '*.html' '*.css' | xargs -0 wc -l | tail -n 1
 103474 total
$ git diff --shortstat
 367 files changed, 70988 insertions(+), 55286 deletions(-)
```

Essentially every file is modified, as should be expected. Some tools
will choke when trying to show a diff for a commit this large (e.g.,
[vim-fugitive]). There’s a valid argument that we shouldn’t design
around those broken tools. But Prettier also supports an incremental
per-file option:

```
  --require-pragma         Require either '@prettier' or '@format' to be present in the file's first docblock comment
                           in order for it to be formatted.
                           Defaults to false.
```

This would enable us to reformat fewer files at a time, while still
protecting against backsliding. We could adopt a soft policy that we
reformat files before modifying them, for instance.

Which should we choose?

## Steps

Here’s what we need to do to make this happen.

  - [x] Make sure that all code is syntactically valid, so that Prettier
    can parse it. Done in #2423.
  - [ ] Check in a `.prettierconfig`. (We like [this one][pc].)
  - Either:
      - [ ] Add `prettier -l 'tensorboard/**/*.'{js,ts,html,css}` to CI,
            and reformat all code.
  - Or:
      - [ ] Add `prettier --require-pragma -l 'tensorboard/**/*.'{js,ts,html,css}`
        to CI.
      - [ ] Adopt a policy of adding `@format` to all new files.
      - [ ] Adopt a soft policy of generally formatting files, modules,
        or packages (choose one?) before working on them.
  - [ ] Update `CONTRIBUTING.md` to show how to configure format-on-save.

[black]: https://github.com/psf/black
[pc]: https://github.com/sourcecred/sourcecred/blob/d5a1ca30b4d1551fdc3bf81b488aa8ae4a37c1d3/.prettierrc.json
[prettier]: https://prettier.io/playground/
[vim-fugitive]: https://github.com/tpope/vim-fugitive

**Comment by wchargin** [comment]
Open for comments.

**Comment by caisq** [comment]
Just to be clear, this will only affect the following four file types:
- html
- css
- js
- ts
Is that correct?

If so, LGTM modulo detailed `.prettierconfig` content. We can discuss and tweak the config once the infrastructure is in place.

The Python style should be separate from prettier IMO. We already have pylint.

**Comment by wchargin** [comment]
> Just to be clear, this will only affect the following four file types:
> 
> * html
> * css
> * js
> * ts
>   Is that correct?

Yes. Prettier can also format JSON, but our JSON data tends to be
autogenerated and we wouldn’t want to format it anyway.

> The Python style should be separate from prettier IMO.

Prettier does not (cannot) format Python, so there you go.

> We already have pylint. 

We don’t actually run pylint. We run `flake8`, but [only for the
following checks][flake8-checks]:

  - E901: `SyntaxError` or `IndentationError`
  - E999: failed to parse AST
  - F821: undefined name
  - F822: undefined name in `__all__`
  - F823: local variable referenced before assignment

That is, we use `flake8` as a correctness formatter, not a style
formatter.

The analogous style formatter for Python would be [Black][black].

[black]: https://github.com/psf/black
[flake8-checks]: https://github.com/tensorflow/tensorboard/blob/a13d9310b41438ff9d60edef61adeee9158a6b87/.travis.yml#L77

**Comment by stephanwlee** [comment]
For the same reason explained in the Why section, did we consider using Black?

**Comment by wchargin** [comment]
> For the same reason explained in the Why section, did we consider
> using Black?

Good question. I did consider it.

Black is even more uncompromising than Prettier. It is entirely
unconfigurable except for max line length, and in particular it always
uses 4-space indentation. This is standard in the Python community (cf.
[PEP 8]), but not at Google—so much so that TensorFlow switched their
style guide entirely from Google style to “PEP 8 but with 2-space
indentation” to avoid a massive diff.

By contrast, Prettier’s default shiftwidth is 2 spaces, and this is also
configurable.

Black [will not support 2-space indentation][no-2]. I’d assumed that
this would be a dealbreaker (though I don’t personally mind either way,
except for the obvious mega-diff, which I do care about).

There is [a fork that uses 2-space indentation][black-2spaces], so we
could in principle use that.

[PEP 8]: https://www.python.org/dev/peps/pep-0008/
[black-2spaces]: https://github.com/desbma/black-2spaces
[no-2]: https://github.com/psf/black/issues/378

**Comment by nfelt** [comment]
I'm generally in favor of opinionated autoformatters, but we should strive to remain compatible with Google's style guides.  Does Prettier's output meet the relevant style guides ([JS](https://google.github.io/styleguide/jsguide.html), [HTML/CSS](https://google.github.io/styleguide/htmlcssguide.html))?

It looks like for TypeScript Google actually publishes a linter/formatter that wraps Prettier, though it uses a slightly different config: https://github.com/google/gts/blob/master/prettier.config.js

As for adoption, my preference would be to either do a single giant commit or to batch it into several commits rebase-and-merged at the same time to keep the diff size down.  I think if we tried to do a gradual rollout over time it'd be hard to enforce (since we don't have a great way, that I know of, to enforce checks only against modified files) and we would not really get the benefit of uniform style everywhere, and when checking blame we'd have to consider multiple different autoformat meteorite events in different geological zones of the codebase rather than just one giant meteorite.

Re: Black, given that we develop in the open and Google's public python style guide actually does endorse 4 spaces, I think it isn't a dealbreaker (we would essentially switch from TF Python style to Google Public Python style), so I'm personally open to adopting it too (just not having to worry about what quote format is in use in the file would be a big win).

One other question is how long it takes to run the formatters over the codebase - flake8 is about 10 seconds.  Ideally we'd avoid adding too much extra PR test latency.

**Comment by wchargin** [comment]
It sounds like people are generally open to this change. I’ll open a PR
with the changes, not to immediately merge, but so that people can see
the style for themselves.

> batch it into several commits rebase-and-merged at the same time to
> keep the diff size down

sgtm

> One other question is how long it takes to run the formatters over the
> codebase - flake8 is about 10 seconds. Ideally we'd avoid adding too
> much extra PR test latency.

Agreed. Prettier takes about 11 seconds. If we like, we can probably
arrange to perform our lint checks in a separate CI “job” so that they
run in parallel and only once. Our current `before_script` block takes
about 16 seconds total, so it might not be worth it.

> we don't have a great way, that I know of, to enforce checks only
> against modified files

Travis runs on a merge commit, so

```
$ git diff-tree --name-only -r -z HEAD^0..HEAD^1 '*.'{js,ts,html,css}
```

should actually do it for PRs… but I expect that we don’t want to put
much effort into maintaining even that. So, agreed.

> Does Prettier's output meet the relevant style guides

tl;dr: mostly, except continuations are +2 rather than +4

<details>
JavaScript and TypeScript: The only major difference that I see is that
continuations are indented by one tab stop rather than two: e.g.,

```javascript
// Google (“most common” wrapping pattern, not explicitly required)
doSomething(
    veryDescriptiveArgumentNumberOne,
    veryDescriptiveArgumentTwo,
    tableModelEventHandlerProxy,
    artichokeDescriptorAdapterIterator) {
  foo();
  // ^ This line is clearly part of the block because it is unindented
  // compared to the preceding line.
}
```

```javascript
// Prettier
doSomething(
  veryDescriptiveArgumentNumberOne,
  veryDescriptiveArgumentTwo,
  tableModelEventHandlerProxy,
  artichokeDescriptorAdapterIterator,
) {
  foo();
  // ^ This line is clearly part of the block because the block begins
  // directly above it, and it is indented compared to the preceding
  // line.
}
```

Note that Prettier’s version is nicer on the blame in the same way that
trailing commas are.

Google style says, “do not mix quoted and unquoted keys” in object
literals. Prettier agrees in cases where keys _can_ be unquoted. When
keys _must_ be quoted (e.g., `{"a-b": c}`), Prettier defaults to quoting
keys only when necessary. The `--quote-props consistent` option changes
this. (Prettier’s default is nicer on the diff, because adding `"a-b"`
as a key to an object with ten unquoted keys doesn’t force all of the
existing keys to change. But I don’t object either way.)

Prettier formats union types like `string | number`, while Closure types
are traditionally written `string|number`. Google style says to include
spaces on both sides of any binary or ternary operator, and “`|`” is
most definitely a binary operator on types, so Prettier is correct and
traditional Closure style is wrong. :-) Also, our current codebase uses
Prettier style overwhelmingly:

```
$ git grep -o '[ ]|[ ]' '*.ts' | wc -l
331
$ git grep -o '[^ ]|[^ ]' '*.ts' | wc -l
27
```

HTML: Same difference with continuations, as above: e.g., Googlers
traditionally write

```html
<!-- Google -->
<my-long-element
  with-some-attribute="one"
  and-another="two"
  and-a-third>
</my-long-element>
```

(sometimes with the closing tag even on the previous line!) where
Prettier would write 

```html
<!-- Prettier -->
<my-long-element
  with-some-attribute="one"
  and-another="two"
  and-a-third
></my-long-element>
```

Same trailing-comma benefit here.

CSS: Generally aligned (one selector per line, always semicolons, etc.).
Prettier is actually relatively unopinionated here: it doesn’t add or
remove quotes in URI values, nor does it permute selectors, nor does it
add or remove blank lines between rules. So afaict it’s “Google style
in, Google style out”.
</details>

**Comment by nfelt** [comment]
Thanks for the extensive research!  I agree none of that sounds like a concern.  SGTM to open a PR.

I assume the thought is to start with just Prettier, and we can see if we like that enough to do this for python as well?

**Comment by wchargin** [comment]
> SGTM to open a PR.

Opened #2473 with just the config and formatting changes: no CI and no
documentation. This is not ready to merge, but feel free to poke around.
(Both “ignore whitespace” and “don’t ignore whitespace” are helpful diff
modes this time, interestingly.)

> I assume the thought is to start with just Prettier, and we can see if
> we like that enough to do this for python as well?

Certainly open to that!

**Comment by stephanwlee** [comment]
@nfelt forgot to add this bit. We gradually will have less JavaScript and more TypeScript. In that world, I expect our modification to look about the same as gts (or just use gts).

**Comment by lakovar** [comment]
I notice that Prettier changes single quotes to double quotes, which is contrary to the internal style guide. Is that okay, or is it worth specifying single quotes in the config?

**Comment by wchargin** [comment]
Hmm; I didn’t realize that the Google JavaScript style guide requires
single quotes. (The Python guide permits either.)

I suppose that we should probably use single quotes throughout, then,
despite my personal preferences to the contrary. Thanks for pointing
this out!

**Comment by wchargin** [comment]
Merged!

For contributors:

  - Run `yarn fix-lint` to reformat all files, which takes about 10
    seconds on my machine. Run `yarn lint` to simply check formatting.

    (You will need to first install [Yarn], and then run `yarn` once in
    the TensorBoard repository to install dependencies.)

  - Consider enabling editor integration for format-on-save. I use the
    official [vim-prettier] plugin with configuration

    ```vim
    let g:prettier#autoformat = 0
    autocmd BufWritePre *.js,*.jsx,*.ts,*.tsx,*.css,*.html PrettierAsync
    ```

    which works quite nicely. It’s asynchronous, so it doesn’t make
    saving slow, either. The buffer generally updates about half a
    second after you save the file.

[Yarn]: https://yarnpkg.com/en/
[vim-prettier]: https://github.com/prettier/vim-prettier
